controller/login.go

```diff
@@ -1,48 +1,61 @@
-
-	client "github.com/fabric8-services/fabric8-auth/client"
 	"github.com/goadesign/goa"
+	"github.com/satori/go.uuid"
 )
 
-type LoginConfiguration interface {
-	login.Configuration
-	GetKeycloakEndpointAuth(*goa.RequestData) (string, error)
-	GetKeycloakURL() string
-	GetKeycloakRealm() string
-	GetPublicOauthClientID() string
-	GetServiceAccounts() map[string]configuration.ServiceAccount
-}
-
 // LoginController implements the login resource.
 type LoginController struct {
 	*goa.Controller
-	Auth          login.KeycloakOAuthService
-	TokenManager  token.Manager
-	Configuration LoginConfiguration
+	app application.Application
 }
 
 // NewLoginController creates a login controller.
-func NewLoginController(service *goa.Service, auth *login.KeycloakOAuthProvider, tokenManager token.Manager, configuration LoginConfiguration) *LoginController {
-	return &LoginController{Controller: service.NewController("login"), Auth: auth, TokenManager: tokenManager, Configuration: configuration}
+func NewLoginController(service *goa.Service, app application.Application) *LoginController {
+	return &LoginController{
+		Controller: service.NewController("login"),
+		app:        app}
 }
 
 // Login runs the login action.
 func (c *LoginController) Login(ctx *app.LoginLoginContext) error {
+	// Generate a new unique state value
+	state := uuid.NewV4().String()
 
-	oauthIdentityProvider := login.NewIdentityProvider(c.Configuration)
-	oauthIdentityProvider.RedirectURL = rest.AbsoluteURL(ctx.RequestData, client.LoginLoginPath(), nil)
+	// Get the URL of the callback endpoint, the client will be redirected here after being redirected to the authentication provider
+	callbackURL := rest.AbsoluteURL(ctx.RequestData, client.CallbackLoginPath(), nil)
+
+	// Remove this: once we remove support for offline token completely
+	var scopes []string
 	if ctx.Scope != nil {
-		oauthIdentityProvider.Endpoint.AuthURL = fmt.Sprintf("%s?scope=%s", oauthIdentityProvider.Endpoint.AuthURL, *ctx.Scope) // Offline token
+		scopes = append(scopes, *ctx.Scope)
 	}
 
-	ctx.ResponseData.Header().Set("Cache-Control", "no-cache")
-	return c.Auth.Login(ctx, oauthIdentityProvider, c.Configuration)
+	redirectURL, err := c.app.AuthenticationProviderService().GenerateAuthCodeURL(ctx, ctx.Redirect, ctx.APIClient,
+		&state, scopes, nil, ctx.RequestData.Header.Get("Referer"), callbackURL)
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, err)
+	}
+	ctx.ResponseData.Header().Set("Location", *redirectURL)
+	return ctx.TemporaryRedirect()
+}
+
+func (c *LoginController) Callback(ctx *app.CallbackLoginContext) error {
+	state := ctx.Params.Get("state")
+	code := ctx.Params.Get("code")
+	redirectURL := rest.AbsoluteURL(ctx.RequestData, client.CallbackLoginPath(), nil)
+
+	redirectTo, err := c.app.AuthenticationProviderService().LoginCallback(ctx, state, code, redirectURL)
+	if err != nil {
+		ctx.ResponseData.Header().Set("Location", *redirectTo+"?error="+err.Error())
```

Comment:
AFAICT when testing locally, if `err != nil`, then `ctx.ResponseData.Header().Set("Location", *redirectTo+"?error="+err.Error())` will cause a panic (`invalid memory address or nil pointer dereference`), probably because `redirectTo` is `nil`

Reply:
yes, I'm already working on it ;)